### ISSUE TITLE
Refactor the way resources are injected into the reconcilers

### ISSUE THREAD

**Issue by pierretasci** [tektoncd/pipeline#3305]
### Feature request

The `TaskRun` and `PipelineRun` reconcilers are very dependent on the `resources` package which implements several fetchers. Many of these fetchers are strongly tied to specific implementations and likewise tied to the cluster itself. These assumptions may not always be the case (as Tekton Bundles has shown) so a refactor could be useful. Ideally, we should be rewrite the interfaces to take the entire reconciliation context and produce the required resource.

**Comment by pritidesai** [comment]
Hey @vdemeester I am moving this out of `0.18` (to `0.19`) as per our discussion in the General WG, "moving bundle behind a feature flag give that its still in `alpha`"

**Comment by sbwsg** [comment]
Pushing this one back to 0.22 since it sounds like there are a number of connected pieces that need to be pulled together to get this into good shape.

**Comment by sbwsg** [comment]
Pushing this one back again to 0.23.

**Comment by sbwsg** [comment]
We could spend some effort taking some measurements in our dogfooding clusters to see how a change here could have an impact before we commit to refactoring?

@pierretasci do you have an idea of how you'd like to proceed on this one?  Is it still worth keeping in the milestones?

**Comment by pierretasci** [comment]
I think this is going to need to happen for remote resource resolution anyways, so perhaps we refocus this on what is needed to get that to work

**Comment by sbwsg** [comment]
Makes sense, I will remove this from milestones for now on the basis that remote resolution is still in the design stages.

**Comment by sbwsg** [comment]
/assign

I've started working on this [in a branch](https://github.com/tektoncd/pipeline/compare/main...sbwsg:refactor-ref-resolution). The way I've structured it at the moment, I've introduced a new reconciler whose sole responsibility is resolving. Once the specs have been resolved into the status field of the taskrun / pipelinerun then the "original" reconcilers take over again and actuate as normal.

**Comment by sbwsg** [comment]
One interesting wrinkle during this refactor: our resolution logic is pretty directly tied to the way we update annotations and labels on taskruns/pipelineruns.  As soon as we've resolved the task or pipeline we copy the labels and annotations over to the run. So at the moment the "Resolver Reconciler" I've introduced is responsible not only for resolving resources but also copying those annotations/labels over too.

At some point we'll probably want to change this. I think ideally the resolution layer shouldn't also be responsible for deciding which labels and annotations are copied to runs.

**Comment by mattmoor** [comment]
> That is, reconciliation should not block but reaching out to external services like an image registry is expensive and error prone

I thought I'd chime in here, since I noticed an experiment that's got multiple reconcilers parting on `*Run` status, which is kind of an anti-pattern.

So we have a number of places we've dealt with problems very similar to this in Knative with a fair degree of success.  Two that are top-of-mind for me are:
 * probing the dataplane for kingress readiness,
 * tag to digest resolution.

_Both of these expose idempotent interfaces to the core reconciler, uses a workqueue for, and has a callback that enqueues the original resource when the job is done._  This pattern has served us very well, and I'd definitely recommend it as a way of offloading work from the main reconciler.

---

That said, I think the main thing I'd change with these is to actually use a **child resource** as the API, and leverage a proper reconciler to manage the workqueuing (and OwnerRefs to handle queuing the parent on completion).

It is notable that this is *effectively* the direction that y'all are heading, except instead of reconciling a child resource you are splitting the reconciliation of a single resource across two reconcilers (which I would **strongly** discourage).

I'd be happy to jump on a call to discuss this more.

**Comment by sbwsg** [comment]
This sounds like the [Tekton Resource Request CRD](https://github.com/tektoncd/community/blob/main/teps/0060-remote-resource-resolution.md#introduce-a-tekton-resource-request-crd) alternative captured in the resolution TEP. That proposal introduces a new resource type (child resource?) and reconciler as you describe. Good to know this is a preferred approach.

**Comment by sbwsg** [comment]
/remove-lifecycle stale

justification: this may still happen if TEP-0060 gets approved and implemented.

**Comment by sbwsg** [comment]
In https://github.com/tektoncd/pipeline/pull/4596 we're adding support for pipelineRefs via remote resolution. This makes the resolution bit non-blocking to a PipelineRun's reconcile. However the code still maintains the PipelineRun's existing resolution mechanics alongside the new stuff and could still use a refactor, perhaps if/when remote resolution moves out of alpha. One option would be for every impl, including Tekton's built-in, to be moved behind the [`remote.Resolver` interface](https://github.com/tektoncd/pipeline/blob/main/pkg/remote/resolver.go#L27-L33) that bundles introduced and Pipeline's reconcilers could just speak to that.
